src/MICore/Transports/LocalUnixTerminalTransport.cs

```diff
@@ -33,49 +43,59 @@ public override void InitStreams(LaunchOptions options, out StreamReader reader,
 
             _dbgStdInName = UnixUtilities.MakeFifo(Logger);
             _dbgStdOutName = UnixUtilities.MakeFifo(Logger);
-
-            string watcherDirectory = _dbgStdInName.Substring(0, _dbgStdInName.IndexOf(UnixUtilities.FifoPrefix, StringComparison.OrdinalIgnoreCase));
-            _fifoWatcher = new FileSystemWatcher(watcherDirectory, UnixUtilities.FifoPrefix + "*");
-            _fifoWatcher.Deleted += FifoWatcher_Deleted;
-            _fifoWatcher.EnableRaisingEvents = true;
+            string pidFifo = UnixUtilities.MakeFifo(Logger);
+            string exitFifo = UnixUtilities.MakeFifo(Logger);
 
             // Setup the streams on the fifos as soon as possible.
             FileStream dbgStdInStream = new FileStream(_dbgStdInName, FileMode.Open);
             FileStream dbgStdOutStream = new FileStream(_dbgStdOutName, FileMode.Open);
+            FileStream pidStream = new FileStream(pidFifo, FileMode.Open);
+            FileStream exitStream = new FileStream(exitFifo, FileMode.Open);
 
             string debuggerCmd = UnixUtilities.GetDebuggerCommand(localOptions);
             string launchDebuggerCommand = UnixUtilities.LaunchLocalDebuggerCommand(
                 debuggeeDir,
-                debuggerCmd,
+                exitFifo,
                 _dbgStdInName,
-                _dbgStdOutName
-                );
+                _dbgStdOutName,
+                pidFifo,
+                debuggerCmd);
 
             TerminalLauncher terminal = TerminalLauncher.MakeTerminal("DebuggerTerminal", launchDebuggerCommand, localOptions.Environment);
             terminal.Launch(debuggeeDir);
 
+            using (StreamReader pidReader = new StreamReader(pidStream, Encoding.UTF8, true, UnixUtilities.StreamBufferSize))
+            {
+                _debuggerPid = int.Parse(pidReader.ReadLine(), CultureInfo.InvariantCulture);
```

Comment:
Maybe a timeout then? Ex: If the pid hasn't returned after 15 seconds something must be wrong.

---
In reply to: [62571354](https://github.com/Microsoft/MIEngine/pull/311#discussion_r62571354) [](ancestors = 62571354)